sotodlib/io/bookbinder2.py

```diff
@@ -0,0 +1,1118 @@
+        will be used to set the `names` in the signal frames. If not provided,
+        names will be taken from the input frames.
+
+    
+    Attributes
+    -----------
+    ancil : AncilProcessor
+        Processor for ancillary data
+    streams : dict
+        Dict of SmurfStreamProcessor objects, keyed by stream_id
+    times : np.ndarray
+        Array of times for all samples in the book
+    frame_idxs : np.ndarray
+        Array of output frame indices for all samples in the book
+    file_idxs : np.ndarray
+        Array of output file indices for all output frames in the book
+    """
+    def __init__(self, book, obsdb, filedb, data_root, readout_ids, outdir,
+                 max_samps_per_frame=50_000, max_file_size=1e9):
+        self.filedb = filedb
+        self.book = book
+        self.data_root = data_root
+        self.hk_root = os.path.join(data_root, 'hk')
+        self.meta_root = os.path.join(data_root, 'smurf')
+        self.hkfiles = get_hk_files(os.path.join(data_root, 'hk'), 
+                                    book.start.timestamp(),
+                                    book.stop.timestamp())
+        self.obsdb = obsdb
+        self.outdir = outdir
+
+        self.max_samps_per_frame = max_samps_per_frame
+        self.max_file_size = max_file_size
+
+        if not os.path.exists(outdir):
+            os.makedirs(outdir)
+
+        logfile = os.path.join(outdir, 'Z_bookbinder_log.txt')
+        self.log = setup_logger(logfile)
+
+        self.ancil = AncilProcessor(self.hkfiles, book.bid, log=self.log)
+        self.streams = {}
+        for obs_id, files in filedb.items():
+            stream_id = '_'.join(obs_id.split('_')[1:-1])
+            self.streams[stream_id] = SmurfStreamProcessor(
+                obs_id, files, book.bid, readout_ids[obs_id], log=self.log
+            )
+
+        self.times = None
+        self.frame_idxs = None
+        self.file_idxs = None
+        self.meta_files = None
+        
+    def preprocess(self):
+        """
+        Runs preprocessing steps for the book. Preprocesses smurf and ancillary
+        data. Creates full list of book-times, the output frame idx for each
+        sample, and the output file idx for each frame.
+        """
+        if self.times is not None:
+            return
+
+        for stream in self.streams.values():
+            stream.preprocess()
+
+        t0 = np.max([s.times[0] for s in self.streams.values()])
+        t1 = np.min([s.times[-1] for s in self.streams.values()])
+        ts, _ = fill_time_gaps(stream.times)
+        m = (t0 < ts) & (ts < t1)
+        ts = ts[m]
+
+        self.ancil.preprocess()
+
+        # Divide up frames
+        frame_splits = find_frame_splits(self.ancil)
+        if frame_splits is None:
+            frame_idxs = np.arange(len(ts)) // self.max_samps_per_frame
+        else:
+            frame_idxs = np.digitize(ts, frame_splits)
+            frame_idxs -= frame_idxs[0]
```

Comment:
Does this handle frames whose length exceeds `self.max_samps_per_frame`?

Reply:
No... I guess we should do that here, though makes the logic a bit more complicated.